core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java

```diff
@@ -66,6 +103,138 @@ public void checkRead(Object key) {
+         log.tracef("Added partially committed (1PC) transaction %s", globalTransaction);
+      }
+      partialTransactions.put(globalTransaction, new Commit1PCTransactionInfo(globalTransaction, affectedNodes,
+                                                                              lockedKeys, modifications));
+      return true;
+   }
+
+   @Override
+   public boolean isTransactionPartiallyRolledBack(GlobalTransaction globalTransaction) {
+      TransactionInfo transactionInfo = partialTransactions.get(globalTransaction);
+      if (trace) {
+         log.tracef("Can release resources for transaction %s. Transaction info=%s", globalTransaction, transactionInfo);
+      }
+      return transactionInfo == null || transactionInfo.isRolledBack(); //if we are going to commit, we can't release the resources yet
+   }
+
+   @Override
+   public Collection<GlobalTransaction> getPartialTransactions() {
+      return Collections.unmodifiableCollection(partialTransactions.keySet());
+   }
+
+   @Override
+   public boolean canRollbackTransactionAfterOriginatorLeave(GlobalTransaction globalTransaction) {
+      boolean canRollback = availabilityMode == AvailabilityMode.AVAILABLE &&
+            !getLastStableTopology().getActualMembers().contains(globalTransaction.getAddress());
+      if (trace) {
+         log.tracef("Can rollback transaction? %s", canRollback);
+      }
+      return canRollback;
+   }
+
+   @Override
+   public void onTopologyUpdate(CacheTopology cacheTopology) {
+      boolean isStable = isTopologyStable(cacheTopology);
+      if (trace) {
+         log.tracef("On stable topology update. Has pending tx? %b. Is stable? %b. topology=%s",
+                    !partialTransactions.isEmpty(), isStable, cacheTopology);
+      }
+      if (isStable) {
+         if (!partialTransactions.isEmpty()) {
+            for (TransactionInfo transactionInfo : partialTransactions.values()) {
+               completeTransaction(transactionInfo, cacheTopology);
+            }
+         }
+      }
+   }
+
+   private void completeTransaction(final TransactionInfo transactionInfo, CacheTopology cacheTopology) {
+      final Collection<Address> affectedNodes = transactionInfo.getCommitNodes(cacheTopology);
+      NotifyingNotifiableFuture<Map<Address, Response>> notifyingFuture = new NotifyingFutureImpl<>();
+      notifyingFuture.attachListener(new FutureListener<Map<Address, Response>>() {
+         @Override
+         public void futureDone(Future<Map<Address, Response>> future) {
+            final GlobalTransaction globalTransaction = transactionInfo.getGlobalTransaction();
+            if (trace) {
+               log.tracef("Future done for transaction %s", globalTransaction);
+            }
+            try {
+               future.get();
+            } catch (Throwable e) {
+               if (trace) {
+                  log.tracef(e, "Exception for transaction %s", globalTransaction);
+               }
+               return;
```

Comment:
And wait for another stable topology to retry? Maybe we could add an explicit check for SuspectException/CacheNotFoundResponse, like in TxDistributionInterceptor, and log an error for unexpected exceptions?